File: laboratorio/sysmex_hemograma.py
"""
Compatibilidad con conversión Sysmex hemograma (delega en entrada_resultados).

Preferir ``laboratorio.entrada_resultados`` con configuración del catálogo TipoExamen.
"""
from __future__ import annotations

from decimal import Decimal
from typing import Any

from laboratorio.catalogo_entrada_default import ENTRADA_DEFAULTS_POR_CODIGO
from laboratorio.entrada_resultados import (
    EntradaRule,
    aplicar_valor_ticket_si_corresponde,
    convert_ticket_entry,
    format_informe_entrada,
    get_entrada_rule,
)


class _CodigoTipoExamen:
    """Proxy mínimo para reutilizar reglas legacy por código."""

    def __init__(self, codigo: str):
        self.codigo = codigo
        self.modo_entrada = "ESTANDAR"


SYSMEX_HEMOGRAMA_UNIDADES: dict[str, str] = {
    "LEU": "/mm³",
    "PLAQ": "/mm³",
    "HEMATIES": "mill/mm³",
    "HGB": "g/dL",
    "HTO": "%",
    "RDW": "%",
    "NEUT_CAY": "%",
    "NEUT_SEG": "%",
    "EOS": "%",
    "BAS": "%",
    "LINF": "%",
    "MONO": "%",
}


def _rules_dict() -> dict[str, EntradaRule]:
    out: dict[str, EntradaRule] = {}
    for codigo, row in ENTRADA_DEFAULTS_POR_CODIGO.items():
        modo, dec, mult, fmt = row
        out[codigo] = EntradaRule(int(dec), Decimal(mult), fmt, modo)
    return out


SYSMEX_HEMOGRAMA_RULES = _rules_dict()


def get_sysmex_unidad(codigo: str | None) -> str:
    if not codigo:
        return ""
    return SYSMEX_HEMOGRAMA_UNIDADES.get(codigo.strip().upper(), "")


def is_sysmex_hemograma_codigo(codigo: str | None) -> bool:
    if not codigo:
        return False
    return codigo.strip().upper() in ENTRADA_DEFAULTS_POR_CODIGO


def convert_sysmex_entry(codigo: str, raw_entry: Any) -> dict[str, Any] | None:
    return convert_ticket_entry(_CodigoTipoExamen(codigo), raw_entry)


def format_sysmex_informe(ticket_value: Decimal, display: str) -> str:
    return format_informe_entrada(ticket_value, display)


def aplicar_valor_sysmex_si_corresponde(resultado, tipo_examen, item: dict[str, Any]) -> bool:
    return aplicar_valor_ticket_si_corresponde(resultado, tipo_examen, item)


def get_sysmex_rule(codigo: str) -> EntradaRule | None:
    return get_entrada_rule(_CodigoTipoExamen(codigo))

File: frontend/src/utils/sysmexHemograma.ts
/**
 * Compatibilidad con código/tests que usan conversión Sysmex por código.
 * La lógica vive en entradaResultados.ts (configuración por catálogo TipoExamen).
 */
import type { LimsTipoExamen } from '../types/lims';
import {
  convertTicketEntry,
  entryFromStored,
  previewTicketInforme,
  formatInformeEntrada,
  usesTicketEntry,
  isFormulaPercent,
  computeFormulaProgress,
  formatFormulaProgressLabel,
  getEntradaRule,
  type TicketConversionResult,
  type FormatoInformeEntrada,
} from './entradaResultados';

export type SysmexDisplayKind = FormatoInformeEntrada;
export type SysmexConversionResult = TicketConversionResult;

export {
  formatInformeEntrada as formatSysmexInforme,
  computeFormulaProgress as computeFormulaLeucocitariaProgress,
  formatFormulaProgressLabel,
};

export const SYSMEX_HEMOGRAMA_UNIDADES: Record<string, string> = {
  LEU: '/mm³',
  PLAQ: '/mm³',
  HEMATIES: 'mill/mm³',
  HGB: 'g/dL',
  HTO: '%',
  RDW: '%',
  NEUT_CAY: '%',
  NEUT_SEG: '%',
  EOS: '%',
  BAS: '%',
  LINF: '%',
  MONO: '%',
};

export const FORMULA_LEUCOCITARIA_CODIGOS = [
  'NEUT_CAY',
  'NEUT_SEG',
  'EOS',
  'BAS',
  'LINF',
  'MONO',
] as const;

function teFromCodigo(codigo: string): LimsTipoExamen {
  return { id: 0, codigo, nombre: codigo, tipo_muestra_requerida: 0, modo_entrada: 'ESTANDAR' };
}

export function getSysmexUnidad(codigo: string | undefined | null): string {
  if (!codigo) return '';
  return SYSMEX_HEMOGRAMA_UNIDADES[codigo.trim().toUpperCase()] ?? '';
}

export function isSysmexHemogramaCodigo(codigo: string | undefined | null): boolean {
  return usesTicketEntry(codigo ? teFromCodigo(codigo) : null, codigo);
}

export function isSysmexFormulaCodigo(codigo: string | undefined | null): boolean {
  return isFormulaPercent(codigo ? teFromCodigo(codigo) : null, codigo);
}

export function convertSysmexEntry(codigo: string, rawEntry: string): TicketConversionResult | null {
  return convertTicketEntry(teFromCodigo(codigo), rawEntry, codigo);
}

export function sysmexEntryFromStored(
  codigo: string,
  valorNumerico: number | string | null | undefined
): string {
  return entryFromStored(teFromCodigo(codigo), valorNumerico, codigo);
}

export function previewSysmexInforme(codigo: string, rawEntry: string): string | null {
  return previewTicketInforme(teFromCodigo(codigo), rawEntry, codigo);
}

/** @deprecated Solo tests legacy que inspeccionan reglas por código */
export const SYSMEX_HEMOGRAMA_RULES = Object.fromEntries(
  FORMULA_LEUCOCITARIA_CODIGOS.map((c) => [c, getEntradaRule(teFromCodigo(c), c)])
);
